server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -171,6 +181,119 @@ abstract class AbstractSearchAsyncAction<Result extends SearchPhaseResult> exten
         this.indexRoutings = indexRoutings;
         this.results = resultConsumer;
         this.clusters = clusters;
+        if (searchListenersList != null && !(searchListenersList.isEmpty())) {
+            this.searchListenersList = searchListenersList;
+            this.searchRequestOperationsListener = new SearchRequestOperationsListener.CompositeListener(this.searchListenersList, logger);
+            instantiateStartMap();
+            instantiateEndMap();
+            instantiateFailMap();
+        }
+    }
+
+    private void instantiateStartMap() {
```

Comment:
Looks like we are trying to build a state machine?

Reply:
You could avoid a map altogether and invoke the listener directly

Reply:
This is not a state machine as such. It is just a simple map for respective start/end phase which stores ```phaseName -> runnable(which is essentially to invoke desired listener method)```. 

This was done to avoid having too many if else condition(example below) and to make it easy to add more phases later.
```
// Using map to avoid below.
startPhase() {
  if (phase is FetchPhase) {
       // invoke fetch start phase listener method
  } else if (phase is QueryPhase) {
      // invoke query start phase listener method
  } // .. and so on
}
```

Let me know if this make sense or I understood you correctly.